README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.
-->

A Flutter plugin for showing custom items in horizontal or vertical scrollable ListView with item snapping functionality.

As soon as the user scrolls through these items, it automatically stops perfectly on an item to get a good user experience. It uses only ScrollView and ScrollController to get this behavior, PageView is not being used here.

<table>
<tr>
<td><img src="https://github.com/Aayush-Thacker/scroll_snap_widgets/blob/main/files/horizontal_recording_less_size.gif?raw=true" alt="Horizontal Items Recording" width="300" height="667"></td>
<td><img src="https://github.com/Aayush-Thacker/scroll_snap_widgets/blob/main/files/vertical_recording_less_size.gif?raw=true" alt="Vertical Items Recording" width="300" height="667"></td>
</tr>
</table>

## Features

This plugin consists of one widget called ScrollSnapWidgets that gives the snapping functionality automatically. It takes a few required properties and is fully customizable with custom items.

There is also a controller class called ScrollSnapWidgetsController having multiple functions with customizations:
+ isFirst() and isLast() to check if the user is at the first or last item in the scrollable area.
+ animateNext(), animatePrevious(), jumpNext(), jumpPrevious() functions help to programitically go through items.
+ animateStart(), animateEnd(), jumpStart(), jumpEnd() functions help to go directly to the starting or ending point programmatically.
+ The dispose() method is also there for resource cleanup.

## Getting started

This widget requires items with the same size of items to work, where size can be height or width depending on the scroll direction. 

When Scroll Direction is *Horizontal*,
itemSize = Width
        &
widgetSize = Height

When Scroll Direction is *Vertical*,
itemSize = Height
        &
widgetSize = Width

If the above two values are provided properly, the package will work perfectly. Also do remember to use the same height or width as the itemSize in your itemBuilder widget.


## Usage
The example on GitHub will result in this kind of output:

<table>
<tr>
<td><img src="https://raw.githubusercontent.com/Aayush-Thacker/scroll_snap_widgets/main/files/home_page_ss.png" alt="Home Page Screenshot" width="300" height="667"></td>
<td><img src="https://github.com/Aayush-Thacker/scroll_snap_widgets/blob/main/files/horizontal_ss_less_size.jpg?raw=true" alt="Horizontal Items Screenshot" width="300" height="667"></td>
<td><img src="https://github.com/Aayush-Thacker/scroll_snap_widgets/blob/main/files/vertical_ss_less_size.jpg?raw=true" alt="Vertical Items Screenshot" width="300" height="667"></td>
</tr>
</table>

Here is a simple example of ScrollSnapWidget with items scrollable Horizontally:

```dart
ScrollSnapWidgets(
    //instance of ScrollSnapWidgetsController
    controller: _controller,
    //total height of the widget because its horizontal
    widgetSize: 300,
    //horizontally scrollable snapping items
    scrollDirection: Axis.horizontal,
    //width of each item
    itemSize: 150,
    //total number of items
    itemCount: Data.allItems.length
    //item widget builder
    itemBuilder: (itemIndex) => SimpleItemCard(
        widgth: 150,
        item: Data.allItems[itemIndex],
    ),
),
```
Then functions can be used on different events to programmatically go through items:

```dart
//for size pass the same double value as itemSize while initialising the controller
final ScrollSnapWidgetsController _controller =
      ScrollSnapWidgetsController(size: 150);

_controller.animateNext(duration = const Duration(milliseconds: 500), curve = Curves.easeOut);
_controller.animatePrevious();

_controller.jumpStart();
_controller.jumpEnd();
```


## Additional information

This is my first ever published package, hope it helps. If there are any issues please post them on GitHub issues section. Also feel free to contribute new features, and bug fixes from you are always appreciated. Thank you.


CHANGELOG.md
## 0.0.7

Documentation was improved, and comments were added to the code.

## 0.0.1

Initial release with a widget with snapping items implementation that can be used for both horizontal and vertically scrollable items. The scroll physics is designed to stop on items automatically. The initial version also includes a controller to implement next-previous and start-end scrolling functionality programmatically.

test/scroll_snap_widgets_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:scroll_snap_widgets/scroll_snap_widgets.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/data.dart
///a simple data model with title and asset image path as item
class ItemDataModel {
  String title;
  String image;
  ItemDataModel({required this.title, required this.image});
}

///class to store static data
class Data {
  ///list of items containing Anime names and Images from the assets folder
  ///the images are from my wallpaper connection, fellow weebs might like them
  static List<ItemDataModel> allItems = [
    ItemDataModel(title: 'Naruto', image: 'assets/1.jpg'),
    ItemDataModel(title: 'Bleach', image: 'assets/2.jpg'),
    ItemDataModel(title: 'Code Geass', image: 'assets/3.jpg'),
    ItemDataModel(title: 'One Piece', image: 'assets/4.jpg'),
    ItemDataModel(title: 'Bungo Stray Dogs', image: 'assets/5.png'),
    ItemDataModel(title: 'Mob Physico 100', image: 'assets/6.jpg'),
    ItemDataModel(title: 'Chainsaw Man', image: 'assets/7.jpg'),
    ItemDataModel(title: 'Gurren Lagann', image: 'assets/8.jpg'),
    ItemDataModel(title: 'Tokyo Revengers', image: 'assets/9.jpg'),
    ItemDataModel(title: 'Hunter x Hunter', image: 'assets/10.jpg'),
    ItemDataModel(title: 'Arcane', image: 'assets/11.jpg'),
    ItemDataModel(title: 'Violet Evergarden', image: 'assets/13.png'),
  ];
}


example/lib/widgets.dart
import 'package:example/data.dart';
import 'package:flutter/material.dart';

///a simple widget to display an item
class SimpleItemCard extends StatelessWidget {
  final double? height;
  final double? widgth;
  final ItemDataModel item;
  const SimpleItemCard(
      {super.key, this.height, this.widgth, required this.item});

  @override
  Widget build(BuildContext context) {
    ///variable size card
    return SizedBox(
      width: widgth,
      height: height,
      child: Card(
        color: Colors.redAccent,

        ///image can not go out of the card (for rounded corner image)
        clipBehavior: Clip.antiAlias,
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            ///image takes and fill the remaining space
            Expanded(
              child: Image.asset(
                item.image,
                fit: BoxFit.fill,
              ),
            ),

            ///text with small padding
            Padding(
              padding: const EdgeInsets.all(5),
              child: Text(
                item.title,
                style: const TextStyle(
                  color: Colors.white,
                  overflow: TextOverflow.ellipsis,
                  fontWeight: FontWeight.bold,
                  fontSize: 15,
                ),
                textAlign: TextAlign.center,
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/scroll_snap_widgets.dart
import 'package:flutter/material.dart';

/*
A custom made scroll physics to get the snapping effect
this is not fully developed by myself, here is the source of the code from a stackoverflow answer:
https:///stackoverflow.com/questions/60458885/how-to-autocorrect-scroll-position-in-a-listview-flutter

This custom made ScrollPhysics helped me a lot with a project, so I decided to use it to make a cool package that could help others too
*/
class PagingScrollPhysics extends ScrollPhysics {
  final double itemDimension;

  const PagingScrollPhysics(
      {required this.itemDimension, ScrollPhysics? parent})
      : super(parent: parent);

  @override
  PagingScrollPhysics applyTo(ScrollPhysics? ancestor) {
    return PagingScrollPhysics(
        itemDimension: itemDimension, parent: buildParent(ancestor));
  }

  double _getPage(ScrollMetrics position) {
    return position.pixels / itemDimension;
  }

  double _getPixels(double page) {
    return page * itemDimension;
  }

  double _getTargetPixels(ScrollMetrics position, double velocity) {
    double page = _getPage(position);
    if (velocity < -tolerance.velocity) {
      page -= 0.5;
    } else if (velocity > tolerance.velocity) {
      page += 0.5;
    }
    return _getPixels(page.roundToDouble());
  }

  @override
  Simulation? createBallisticSimulation(
      ScrollMetrics position, double velocity) {
    /// If we're out of range and not headed back in range, defer to the parent
    /// ballistics, which should put us back in range at a page boundary.
    if ((velocity <= 0.0 && position.pixels <= position.minScrollExtent) ||
        (velocity >= 0.0 && position.pixels >= position.maxScrollExtent)) {
      return super.createBallisticSimulation(position, velocity);
    }
    final double target = _getTargetPixels(position, velocity);
    if (target != position.pixels) {
      return ScrollSpringSimulation(spring, position.pixels, target, velocity,
          tolerance: tolerance);
    }
    return null;
  }

  @override
  bool get allowImplicitScrolling => false;
}

/*
Controller to control the behaviour of ScrollSnapWidgets
use it's instance with the widget to use all the functionalities of this package
there can be more functions added, if there is any requirement let me know in the GitHub issues
*/
class ScrollSnapWidgetsController {
  ///size should be the same as itemSize (width for horizontal) and (height for vertical)
  double size;
  ScrollSnapWidgetsController({
    required this.size,
  });

  ///initilizing the ScrollController to use other functionalities
  ScrollController itemScrollController = ScrollController();

  ///used to check if the user is at the first item
  bool isFirst() {
    if (itemScrollController.position.pixels < size) {
      return true;
    } else {
      return false;
    }
  }

  ///function used to check if the user is on the edge of the scrollview (not end, it cound also be starting edge)
  bool isLast() {
    if (itemScrollController.position.atEdge) {
      return true;
    } else {
      return false;
    }
  }

  ///animate to the start item with custom duration and curve
  bool animateStart({
    ///returns true if action is done, false otherwise
    Duration duration = const Duration(milliseconds: 500),
    Cubic curve = Curves.easeIn,
  }) {
    ///checking if the user is not already at the start
    if (!isFirst()) {
      itemScrollController.animateTo(
          itemScrollController.position.minScrollExtent,
          duration: duration,
          curve: curve);
      return true;
    }
    return false;
  }

  ///animate to the end item with custom duration and curve
  bool animateEnd({
    ///returns true if action is done, false otherwise
    Duration duration = const Duration(milliseconds: 500),
    Cubic curve = Curves.easeOut,
  }) {
    ///checking if the user is on the last edge
    if (!isLast() || isFirst()) {
      itemScrollController.animateTo(
          itemScrollController.position.maxScrollExtent,
          duration: duration,
          curve: curve);
      return true;
    }
    return false;
  }

  ///go directly to the start
  bool jumpStart() {
    ///returns true if action is done, false otherwise
    ///checking if the user is not already at the start
    if (!isFirst()) {
      itemScrollController.jumpTo(
        itemScrollController.position.minScrollExtent,
      );
      return true;
    }
    return false;
  }

  ///go directly to the end
  bool jumpEnd() {
    ///returns true if action is done, false otherwise
    ///checking if the user is on the last edge
    if (!isLast() || isFirst()) {
      itemScrollController.jumpTo(
        itemScrollController.position.maxScrollExtent,
      );
      return true;
    }
    return false;
  }

  ///animate to the next item with custom duration and curve
  bool animateNext({
    ///returns true if action is done, false otherwise
    Duration duration = const Duration(milliseconds: 500),
    Cubic curve = Curves.easeIn,
  }) {
    ///checking if the user is on the last edge
    if (!isLast() || isFirst()) {
      itemScrollController.animateTo(itemScrollController.offset + size,
          duration: duration, curve: curve);
      return true;
    }
    return false;
  }

  ///animate to the previous item with custom duration and curve
  bool animatePrevious({
    ///returns true if action is done, false otherwise
    Duration duration = const Duration(milliseconds: 500),
    Cubic curve = Curves.easeOut,
  }) {
    ///checking if the user is not already at the start
    if (!isFirst()) {
      itemScrollController.animateTo(itemScrollController.offset - size,
          duration: duration, curve: curve);
      return true;
    }
    return false;
  }

  ///go directly to the next item
  bool jumpNext() {
    ///returns true if action is done, false otherwise
    ///checking if the user is on the last edge
    if (!isLast() || isFirst()) {
      itemScrollController.jumpTo(
        itemScrollController.offset + size,
      );
      return true;
    }
    return false;
  }

  ///go directly to previous item
  bool jumpPrevious() {
    ///returns true if action is done, false otherwise
    ///checking if the user is not already at the start
    if (!isFirst()) {
      itemScrollController.jumpTo(
        itemScrollController.offset - size,
      );
      return true;
    }
    return false;
  }

  ///disposing the ScrollController
  void dispose() {
    itemScrollController.dispose();
  }
}

///main widget of the package, usable for both horizontal and vertical items
class ScrollSnapWidgets extends StatelessWidget {
  ///the direction in which these items will be scrollable
  final Axis scrollDirection;

  ///the fixed size of each item (width for Vertical) and (height for Horizontal)
  final double itemSize;

  ///full widget size in the screen (height for Vertical) and (width for Horizontal)
  final double widgetSize;

  ///total number of items in the list
  final int itemCount;

  ///the item widget builder
  final Widget Function(int itemIndex) itemBuilder;

  ///main padding of the scrollable items
  final EdgeInsets padding;

  ///controller to implement next-previous functionality
  final ScrollSnapWidgetsController? controller;

  ///scroll physics of the parent scrollview (if any)
  final ScrollPhysics? parent;

  ///default constructor with all properties
  const ScrollSnapWidgets({
    super.key,
    required this.widgetSize,
    required this.scrollDirection,
    required this.itemSize,
    required this.itemCount,
    required this.itemBuilder,
    this.controller,
    this.parent,
    this.padding = EdgeInsets.zero,
  });

  @override
  Widget build(BuildContext context) {
    /*
    The UI consists of a SizedBox with fixed height or width according to the scrollDirection
    There is an optional padding as it's child
    The main view is a ListView builder having the custom scroll physics and other custom properties
    */
    return SizedBox(
      height: scrollDirection == Axis.horizontal ? widgetSize : null,
      width: scrollDirection == Axis.vertical ? widgetSize : null,
      child: Padding(
        padding: padding,
        child: ListView.builder(
          scrollDirection: scrollDirection,
          physics: PagingScrollPhysics(itemDimension: itemSize, parent: parent),
          controller: controller?.itemScrollController,
          itemCount: itemCount,
          itemBuilder: (context, index) => itemBuilder(index),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:example/data.dart';
import 'package:scroll_snap_widgets/scroll_snap_widgets.dart';
import 'package:example/widgets.dart';
import 'package:flutter/material.dart';

/*
This is an example file for the Flutter package called scroll_snap_widget
this widget allows us to implement accurate auto-stopping on items in a horizontal or vertical scroll view
it gives a unique scrolling experience, it should not be used if you want to implement fast scrolling,
as it prevents more scrolling than the itemSize and brings the user back to the start of the item (the item can be any Widget)

Please find the assets and data.dart and widgets.dart files from the GitHub repository example to run this code perfectly
or clone the project and run the example, if you want to get the exact or similar output
Of course, you are welcome to try it yourself, this example is just a demonstration of its capabilities, you can use it in many ways

Hope it helps, let me know if there are any bugs or any new features ideas in the GitHub issue section.

Created by,
Aayush-Thacker
*/

void main() {
  runApp(const MyApp());
}

///the first app widget to build the MaterialApp
class MyApp extends StatelessWidget {
  const MyApp({super.key});
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Scroll Snap Widgets Example',
      theme: ThemeData(
        primarySwatch: Colors.red,
      ),
      home: const HomePage(),
    );
  }
}

///HomePage of the example with 2 buttons and an appbar
class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      ///appbar with constant title
      appBar: AppBar(
        title: const Text('Scroll Snap Widgets Example'),
        centerTitle: true,
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          ///first button takes to the HorizontalItems screen
          Expanded(
            child: Padding(
              padding: const EdgeInsets.all(8),
              child: Card(
                margin: EdgeInsets.zero,
                color: Colors.redAccent,
                child: InkWell(
                  onTap: () {
                    Navigator.push(
                        context,
                        MaterialPageRoute(
                            builder: (_) => const HorizontalItems()));
                  },
                  child: const Center(
                    child: Text(
                      'Horizontal Items',
                      style: TextStyle(fontSize: 30, color: Colors.white),
                    ),
                  ),
                ),
              ),
            ),
          ),

          ///second button takes to the VerticalItems screen
          Expanded(
            child: Padding(
              padding: const EdgeInsets.only(bottom: 8, right: 8, left: 8),
              child: Card(
                margin: EdgeInsets.zero,
                color: Colors.redAccent,
                child: InkWell(
                  onTap: () {
                    Navigator.push(
                        context,
                        MaterialPageRoute(
                            builder: (_) => const VerticalItems()));
                  },
                  child: const Center(
                    child: Text(
                      'Vertical Items',
                      style: TextStyle(fontSize: 30, color: Colors.white),
                    ),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

///page to saw the use cases of package with horizontal scrollable items
class HorizontalItems extends StatefulWidget {
  const HorizontalItems({super.key});

  @override
  State<HorizontalItems> createState() => _HorizontalItemsState();
}

class _HorizontalItemsState extends State<HorizontalItems> {
  static const double bigItemWidth = 300;

  ///controller for animated example
  final ScrollSnapWidgetsController _controllerAnimate =
      ScrollSnapWidgetsController(size: bigItemWidth);

  ///controller for jumping example
  final ScrollSnapWidgetsController _controllerJump =
      ScrollSnapWidgetsController(size: bigItemWidth);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Horizontal Items'),
        centerTitle: true,
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            const Padding(
              padding: EdgeInsets.all(8.0),
              child: Text(
                'Normall Snapping Scrollable Items',
                style: TextStyle(fontSize: 20),
              ),
            ),

            ///simple use case without controller
            ScrollSnapWidgets(
                padding: EdgeInsets.zero,

                ///widget size is height in this example
                widgetSize: 200,
                scrollDirection: Axis.horizontal,

                ///item size is item's width
                itemSize: 200,
                itemBuilder: (itemIndex) => SimpleItemCard(
                      ///keep same as item size
                      widgth: 200,
                      item: Data.allItems[itemIndex],
                    ),
                itemCount: Data.allItems.length),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Row(
                children: [
                  const Expanded(
                    child: Text(
                      'Animate Next Previous',
                      style: TextStyle(
                          fontSize: 20, overflow: TextOverflow.ellipsis),
                    ),
                  ),
                  IconButton(
                    onPressed: () {
                      ///animate previous with the default curve and duration
                      _controllerAnimate.animatePrevious();
                    },
                    icon: const Icon(Icons.navigate_before),
                  ),
                  IconButton(
                    onPressed: () {
                      ///animaet next with custom curve and duration
                      _controllerAnimate.animateNext(
                          duration: const Duration(seconds: 1),
                          curve: Curves.linearToEaseOut);
                    },
                    icon: const Icon(Icons.navigate_next),
                  ),
                ],
              ),
            ),

            ///example with controller to control the widget's scroll event programetically
            ScrollSnapWidgets(
                controller: _controllerAnimate,
                padding: EdgeInsets.zero,

                ///widget's width
                widgetSize: 300,
                scrollDirection: Axis.horizontal,
                itemSize: bigItemWidth,
                itemBuilder: (itemIndex) => SimpleItemCard(
                      widgth: bigItemWidth,
                      item: Data.allItems[itemIndex],
                    ),
                itemCount: Data.allItems.length),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Row(
                children: [
                  const Expanded(
                    child: Text(
                      'Jump Next Previous',
                      style: TextStyle(
                          fontSize: 20, overflow: TextOverflow.ellipsis),
                    ),
                  ),
                  IconButton(
                    onPressed: () {
                      ///controller function call for previous jumping
                      _controllerJump.jumpPrevious();
                    },
                    icon: const Icon(Icons.navigate_before),
                  ),
                  IconButton(
                    onPressed: () {
                      ///controller function call for next jumping
                      _controllerJump.jumpNext();
                    },
                    icon: const Icon(Icons.navigate_next),
                  ),
                ],
              ),
            ),
            ScrollSnapWidgets(
                controller: _controllerJump,
                padding: EdgeInsets.zero,

                ///width of the widget
                widgetSize: 300,
                scrollDirection: Axis.horizontal,
                itemSize: bigItemWidth,
                itemBuilder: (itemIndex) => SimpleItemCard(
                      widgth: bigItemWidth,
                      item: Data.allItems[itemIndex],
                    ),
                itemCount: Data.allItems.length),
          ],
        ),
      ),
    );
  }

  ///the dispose method of the controller disposes of the scroll controller
  @override
  void dispose() {
    _controllerAnimate.dispose();
    _controllerJump.dispose();
    super.dispose();
  }
}

///example to demonstrate Vertically snapping items that stop autometically when user scrolls like instagram posts
class VerticalItems extends StatefulWidget {
  const VerticalItems({super.key});

  @override
  State<VerticalItems> createState() => _VerticalItemsState();
}

class _VerticalItemsState extends State<VerticalItems> {
  static const double itemHeight = 500;

  ///controller for handling custom scroll event callbacks
  final ScrollSnapWidgetsController _controller =
      ScrollSnapWidgetsController(size: itemHeight);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Vertical Items'),
        centerTitle: true,
      ),
      body: SizedBox(
        height: double.maxFinite,
        width: double.maxFinite,
        child: Stack(
          ///defines the width of the widget
          fit: StackFit.expand,
          children: [
            ///vertical scrollable items with controller example
            ScrollSnapWidgets(
                controller: _controller,

                ///widget height
                widgetSize: double.maxFinite,
                scrollDirection: Axis.vertical,

                ///item height
                itemSize: itemHeight,
                itemBuilder: (itemIndex) => SimpleItemCard(
                      ///same height as above
                      height: itemHeight,
                      item: Data.allItems[itemIndex],
                    ),
                itemCount: Data.allItems.length),
            Align(
              alignment: Alignment.bottomRight,
              child: SizedBox(
                width: 100,
                height: 200,
                child: Card(
                  color: Colors.white.withOpacity(0.6),
                  elevation: 5,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      Expanded(
                        child: InkWell(
                          onTap: () {
                            ///function call to animate to the starting point
                            _controller.animateStart();
                          },
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: const [
                              Icon(
                                Icons.arrow_upward,
                                size: 35,
                              ),
                              Text('Animate Start'),
                            ],
                          ),
                        ),
                      ),
                      Expanded(
                        child: InkWell(
                          onTap: () {
                            ///function call to animate to the ending point with custom duration
                            _controller.animateEnd(
                                duration: const Duration(seconds: 2));
                          },
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: const [
                              Text('Animate End'),
                              Icon(
                                Icons.arrow_downward,
                                size: 35,
                              ),
                            ],
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
            Align(
              alignment: Alignment.bottomLeft,
              child: SizedBox(
                width: 100,
                height: 200,
                child: Card(
                  color: Colors.white.withOpacity(0.6),
                  elevation: 5,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      Expanded(
                        child: InkWell(
                          onTap: () {
                            ///function call for jumping to the starting point
                            _controller.jumpStart();
                          },
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: const [
                              Icon(
                                Icons.arrow_upward,
                                size: 35,
                              ),
                              Text('Jump Start'),
                            ],
                          ),
                        ),
                      ),
                      Expanded(
                        child: InkWell(
                          onTap: () {
                            ///function call for jumping to the ending point
                            _controller.jumpEnd();
                          },
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: const [
                              Text('Jump End'),
                              Icon(
                                Icons.arrow_downward,
                                size: 35,
                              ),
                            ],
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  ///the dispose method of the controller disposes of the scroll controller
  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }
}
